### DISCUSSION TITLE
VSCode is NOT crashing when running a shallow copies program

### DISCUSSION THREAD

**Discussion by AAnthon1985** [microsoft/vscode-cpptools#7267] [Q&A]
Hi,
Learning C++, I am currently looking at the pitfalls when using shallow copy constructors. The program compiles well, and, unfortunately, it also runs well in VSCode, while it should crash. Is this VSCode specific? Running the program in my console CMD, the program crashes. Running the program in an online compiler, I get the feedback of memory violation. Thank you,

```
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::vector;

class Shallow
{
private:
    int *data;
public:
    void set_data(int d)
    {
        *data = d;
    }
    int get_data_value()
    {
        return *data;
    }

    // Constructor
    Shallow(int d);
    // Copy constructor
    Shallow(const Shallow &source);
    //Destructor
    ~Shallow();
};

// Constructor
Shallow::Shallow(int d)
{
    data = new int;
    *data = d;
}

// Copy constructor
Shallow::Shallow(const Shallow &source)
: data(source.data)
{
    cout << "Copy constructor - shallow copy" << endl;
}

// Destructor
Shallow::~Shallow()
{
    delete data;
    cout << "Destructor freeing data" << endl;
}

void display_shallow(Shallow s)
{
    cout << s.get_data_value() << endl;
}

int main()
{
    Shallow obj1(100);
    display_shallow(obj1);

    Shallow obj2(obj1);
    obj2.set_data(1000);
    
    return 0;
}

```

My tasks.json file:
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "cppbuild",
			"label": "C/C++: g++.exe build active file",
			"command": "C:\\MinGW\\bin\\g++.exe",
			"args": [
				"-g",
				"${workspaceFolder}\\*.cpp",
				// "${workspaceFolder}\\src\\*.cpp",
				"-I",
				"${workspaceFolder}\\inc\\",
				"-o",
				"${fileDirname}\\${fileBasenameNoExtension}.exe"
			],
			"options": {
				"cwd": "${workspaceFolder}"
			},
			"problemMatcher": [
				"$gcc"
			],
			"group": {
				"kind": "build",
				"isDefault": true
			},
			"detail": "compiler: C:\\MinGW\\bin\\g++.exe"
		}
	]
}
```

**Comment by sean-mcmanus** [comment]
It crashes for me in VS Code. Are you sure you're running the app until it runs the destructor?

![image](https://user-images.githubusercontent.com/19859882/113318072-ecdbb880-92c4-11eb-93d5-ae304282e1b8.png)

**Reply by AAnthon1985** [reply]
Hi,
thank you for trying it out on your machine/VS Code. How can I verify that I am running the app until it runs the destructor? My procedure is described below. Also, when debugging the code, I can debug all the way until the end of the main function, right until the return 0;
1. Compile the code via CTRL+Shift+B
```
Starting build...
C:\MinGW\bin\g++.exe -g C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy\*.cpp -I C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy\inc\ -o C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy\main.exe
Build finished successfully.
```
2. In the terminal, I simply run the code by writing .\main
```
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy> .\main
Copy constructor - shallow copy
100
Destructor freeing data        
Copy constructor - shallow copy
Destructor freeing data        
Destructor freeing data        
PS C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy> 
```

![Shallow_copy](https://user-images.githubusercontent.com/68788787/113319214-80d37180-9311-11eb-849e-a38aed92f152.png)

**Reply by sean-mcmanus** [reply]
You have to step beyond the return 0, because the destructor isn't run until the local object gets out of scope (after executing the return).

**Reply by AAnthon1985** [reply]
Okay, one step further. Just did that, and I get a "Paused on exception" in my Thread #1.
But... It says something different, and I don't get the nice red "Exception has occurred" just as you have shown in your demo. It seems more that I am trying to single step outside the code, which makes sense. But no exception regarding my destructor. Anything I am doing wrong perhaps in my debugger configuration? My launch.json file:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "g++.exe - Build and debug active file",
            "type": "cppdbg",
            "request": "launch",
            "program": "${fileDirname}\\${fileBasenameNoExtension}.exe",
            "args": [],
            "stopAtEntry": true,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "C:\\MinGW\\bin\\gdb.exe",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            "preLaunchTask": "C/C++: g++.exe build active file"
        }
    ]
}
```

![Shallow_copy](https://user-images.githubusercontent.com/68788787/113323743-9bf4b000-9316-11eb-9b0a-6ab3bc2308e8.png)

**Reply by sean-mcmanus** [reply]
You're hitting some "Error: Unexpected GDB output". I'm using gdb/g++ 10.1. It looks like you're using an older version which could have some bug in the tooling and/or the C runtime. Are you able to set a breakpoint in ~Shallow() on the call to delete and then step over that (which should trigger the crash)?

**Reply by AAnthon1985** [reply]
EDIT:
@sean-mcmanus I made an interesting discovery. I apparently used to use an older version of MinGW. I didn't notice that there exists a MinGW-w64, which runs the version 8.1.0 for GCC, G++ and GDB.
So here is the pattern that I found (with GDB 8.1):
1. Compiling the code works fine, and running it in VSCode (not the debugger, but in the terminal) seems to work fine in terms of not getting an exception.
2. When afterwards running the debugger, I get no exception and I am basically at the same point where I was when I started this discussion. 
3. Restarting my laptop, opening VSCode and running the debugger then causes the exception just like you showed on your demonstration. So I can reproduce your outcome.
4. Going back to the terminal in VSCode and running the program works fine and I am back at step 1. Running the debugger after step 4 does not cause an exception.

Would it be possible to ask you to reproduce my steps 1-4 to see if we obtain the same results? It would be nice if you could confirm my findings.

Thank you,

**Reply by sean-mcmanus** [reply]
How exactly are you running it in the terminal? Your PATH variable needs to have the mingw path in it, e.g. `C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin` . Do you get the cout messages? By "works fine" I assume you mean it prints 4 lines of output -- you wouldn't see any exception message in the terminal or shell, i.e. the app is just silently crashing.

Restarting your computer and VS Code shouldn't cause different behavior. Are you sure you're building/running the same program?

**Reply by AAnthon1985** [reply]
I do get the cout messages, yes. The mingw path is in it, both when compiling and when debugging. I will show you a step-by-step picture procedure, so you can hopefully better follow what I am doing and the output I receive:

### Step 1: Laptop started, VSCode started, Project opened
![Step_1](https://user-images.githubusercontent.com/68788787/113390692-845b0d00-9392-11eb-8f30-7f0e98cc58bb.png)

### Step 2: Compile the program: CTRL+Shift+B
![Step_2](https://user-images.githubusercontent.com/68788787/113390794-b7050580-9392-11eb-9726-e9b958ccafcc.png)

### Step 3: I press the key "+" to have a new terminal in my current working directory 
![Step_3](https://user-images.githubusercontent.com/68788787/113391226-7b1e7000-9393-11eb-82eb-1b0ceaee46ac.png)

### Step 4: I execute the program by typing `.\main`
![Step_4](https://user-images.githubusercontent.com/68788787/113391348-b456e000-9393-11eb-8b4f-09be8bc894f3.png)

### Step 5: The program is executed and I am back in my terminal command line
![Step_5](https://user-images.githubusercontent.com/68788787/113391482-dfd9ca80-9393-11eb-9057-3b0d4500c3d6.png)

### Step 6: I press F5 to start the debugger
![Step_6](https://user-images.githubusercontent.com/68788787/113391917-9342bf00-9394-11eb-83ba-94c4aa10df16.png)

### Step 7: I jump to the breakpoint when leaving the destructor for the 1st time
![Step_7](https://user-images.githubusercontent.com/68788787/113392257-1c59f600-9395-11eb-82c5-a3bb2b2082ce.png)

### Step 8: I press F5 to jump to the breakpoint when leaving the destructor for the 2nd time
![Step_8](https://user-images.githubusercontent.com/68788787/113392498-84104100-9395-11eb-84a2-14affb228b32.png)

### Step 9: I press F5 to jump to the breakpoint when leaving the destructor for the 3rd time
![Step_9](https://user-images.githubusercontent.com/68788787/113392572-a3a76980-9395-11eb-80e1-7d00939e6746.png)

### Step 10: I press F5 to continue the breakpoint. Debugging finished
![Step_10](https://user-images.githubusercontent.com/68788787/113392725-dcdfd980-9395-11eb-99fb-3dc9f0c5036f.png)

**Reply by sean-mcmanus** [reply]
Some thing seems wrong (or different) with your executable. I don't know what, since I also tried with the mingw-w64 8.1 g++ compiler and I get
```
Copy constructor - shallow copy
100
Destructor freeing data
Copy constructor - shallow copy
```
i.e. crashes afterwards.

My guess is your C:\WINDOWS\System32\msvcrt.dll may be old. I'm on Windows 10 (msvcrt.dll is dated October 2020). The debugger doesn't seem to be a factor. Might be worth trying "C:\Windows\System32\appverif.exe" to see if that is able to detect the issue -- after adding the exe and running it, a log should be generated that will report if it crashed or not.

**Reply by AAnthon1985** [reply]
Okay, so now I am getting confused. First of all, my msvcrt.dll is dated 11/02/2021, so pretty new. Also, perhaps worth to mention, is that my machine is pretty new (less than 6 weeks old). I think/thought I have all the windows updates. At least there are no updates shown to be installed on my Windows update. 
Anyways...
Once I found out that there was no appverif.exe in my System32 folder, I googled it and found out that I did not have the Windows 10 SDK (and hence no Application Verifier) on my machine. So I downloaded Windows 10 SDK, and tried to debug the main.exe file while logging the execution with appverif.exe.
I crashed during debug.

So here comes the mysterious part:
Whenever I am now compiling and running my code in the VSCode terminal (refer to Step 4 and Step 5 from above), I now get the following output:
```
PS C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy> .\main
Copy constructor - shallow copy
100
Destructor freeing data        
Copy constructor - shallow copy
PS C:\Users\Alexander\Documents\GitHub\CPP_Projects\01_beginning_cpp_programming_course\148_Shallow_copy> 
```

Look at that: The program ends without having the last two destructors executed.

In fact, I'm getting the same output as you do.

Not sure what changed on my machine after I installed Windows 10 SDK. Do you perhaps have an explanation?

Cheers,

**Reply by sean-mcmanus** [reply]
Yeah, your C runtime that was being used by mingw must have been some old version (not sure where from). The Windows SDK's C runtime is at a path like `C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt` and `C:\Program Files (x86)\Windows Kits\10\Lib\10.0.19041.0\ucrt\x86`.

**Reply by AAnthon1985** [reply]
Thank you, I'm getting closer here, but there are still some things not entirely understandable for me.

I basically blindly followed the instructions from here:
https://code.visualstudio.com/docs/cpp/config-mingw
Which forwarded me to the following website to download Mingw-w64
https://sourceforge.net/projects/mingw-w64/

So I downloaded and installed mingw-w64 and continued following the instruction from here:
https://code.visualstudio.com/docs/cpp/config-mingw

I did nothing more, and I did nothing less. With that in mind, I am confused that I was running into these problems with the shallow copying.

Let me ask the other way around:
Is Windows 10 SDK a requirement in order to run Mingw-w64 properly, or is something wrong my Mingw-w64? You mentioned that you used gdb/c++ 10.1:

Where did you get that one from? I couldn't see any option as of which gdb/c++ to use. 

The reason why I am asking:
What is it that I need to be aware of in the future when programming in C++ to make sure that things run properly. Clearly, installing Mingw-w64 and VSCode alone is not enough (in my case at least).

Thank you,

**Reply by sean-mcmanus** [reply]
Those docs needs to be updated. The version of mingw described there is old -- see https://github.com/microsoft/vscode-cpptools/issues/6536 . I recommend using msys2 to get gcc 10.2 compilers (or using WSL2 with Ubuntu 20). I'm not entirely sure which components of mingw or msys2 rely on installed Windows 10 SDK (my guess is it might bundle an older version if a newer version isn't found on your system). I always have the Windows 10 SDK installed as part of my VS install.

**Reply by AAnthon1985** [reply]
@sean-mcmanus That's great to hear. Thank you very much (once again), for your help and support. Although with SDK and G++ 8.1.0 it works fine as expected, it's good to know that there is a G++10.2 and that the docs should be updated.

Good job. Issue is resolved and answered from my side here.

Cheers,
